docs/architecture.md

```diff
@@ -0,0 +1,331 @@
+or proprietary async runtime requirements out of any public interfaces (No channels!).
+
+## Domains:
+
+```mermaid
+flowchart TB
+    subgraph Fuel Core
+        direction TB
+        subgraph POAS[PoA Service]
+            POAD{{Domain}}
+            SYNCPOA[Synchronizer Adapter]
+        end
+        subgraph BPS[Block Producer Service]
+            BPD{{Domain}}
+            POABP[PoA Adapter]
+        end
+        subgraph BIS[Block Importer Service]
+            BID{{Domain}}
+            SYNCBI["Sync Adapter"]
+            POABI["PoA Adapter"]
+        end
+        subgraph SS[Synchronizer Service]
+            SSD{{Domain}}
+        end
+        subgraph P2PS[P2P Service]
+            P2PD{{Domain}}
+            POAP2P[PoA Adapter]
+            SYNCP2P[Sync Adapter]
+            TXP2P[Transaction Pool Adapter]
+        end
+        subgraph TXS[Transaction Pool Service]
+            TXD{{Domain}}
+            BPTXS[Block Producer Adapter]
+            BITXS[Block Importer Adapter]
+        end
+        subgraph RS[Relayer Service]
+            RSD{{Domain}}
+            POAR[POA Adapter]
+            BPR[Block Producer Adapter]
+            BIR[Block Importer Adapter]
+        end
+        subgraph ES[Executor Service]
+            ESD{{Domain}}
+            BPEXE[Block Producer Adapter]
+            BIEXE[Block Importer Adapter]
+        end
+        BID & BPD --> ES & RS & TXS
+        SSD --> POAS & P2PS & BIS
+        TXD --> P2PS
+        POAD --> P2PS & RS & BPS & BIS
+        
+    end
+```
+
+### (PoA) Consensus Module
+
+#### Ports: fuel_poa_consensus::ports
+```rust
+trait BlockProducer {
+    // used by the PoA node to produce a new block and immediately save the resultant block state
+    async fn produce_and_store_block(height: BlockHeight) -> Result<FuelBlock>;
+}
+
+trait BlockImporter {
+    // used by non-poa nodes in response to blocks proposed via p2p
+    async fn commit(sealed_block: SealedFuelBlock) -> Result<()>;
+}
+
+trait Relayer {
+    // wait until the relayer eth node is synced
+    async fn await_synced() -> Result<()>;
+    // wait until the relayer reaches a desired da_height
+    async fn await_da_height(da_height: DaHeight) -> Result<()>;
+    // get the consensus key associated with a da_height
+    async fn get_poa_key(da_height: DaHeight) -> Result<ConsensusKey>;
+}
+
+trait Database {
+    // get the current fuel chain height
+    fn get_current_height() -> Result<BlockHeight>;
+    // insert consensus information associated with a sealed block that was locally produced
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+trait PeerToPeer {
+    // broadcast a newly produced block to the network
+    async fn broadcast_new_block(block: SealedFuelBlock) -> Result<()>;
+    // receive a newly produced block from the network
+    async fn on_new_block(block: SealedFuelBlock) -> Result<()>;
```

Comment:
I guess you meant the next signature:
```rust
async fn on_new_block() -> Result<SealedFuelBlock>;
```

Why do we need to monitor the block on the CM side, if Synchronizer does the same?

Reply:
ah right, this should be impl of the domain, not the port's adapter.

The intention was to have a callback-style interface, which I guess would really look like this:

```rust
async fn on_new_block<F: Fn(SealedFuelBlock)>(block_callback: F);
```

But this may not have the best semantics either, for instance, there's no good reason to be able to register multiple callbacks from the CM perspective.

Perhaps a stream would be a more straightforward interface to consume:

```rust
fn broadcast_blocks() -> impl Stream<SealedFuelBlock>;
```